algebra-jackson/src/main/java/com/hubspot/algebra/ResultDeserializer.java

```diff
@@ -62,7 +63,15 @@ private static Object deserializeValue(
       // Our version of Jackson doesn't allow custom deserialization of null
       return NullValue.get();
     } else {
-      return objectMapper.readerFor(type).readValue(valueNode);
+      final ObjectReader reader;
+      if (codec instanceof ObjectMapper) {
+        reader = ((ObjectMapper) codec).readerFor(type);
+      } else if (codec instanceof ObjectReader) {
+        reader = ((ObjectReader) codec).forType(type);
+      } else {
+        throw new IllegalArgumentException("Unable to read values using codec with class: " + codec.getClass());
+      }
```

Comment:
What about something like:
`objectCodec.readValue(valueNode.traverse(objectCodec), type);`
I think this would eliminate all the `instanceof` checks?

Reply:
There's no `ObjectCodec#readValue` that takes a `JavaType`.

Reply:
I see `ObjectCodec#readValue(JsonParser, ResolvedType)` where `ResolvedType` is a superclass of `JavaType`

Reply:
Sorry, I should've said `JsonNode` and `JavaType`

Reply:
`valueNode.traverse(objectCodec)` returns a `JsonParser`